cfme/cloud/provider/__init__.py

```diff
@@ -309,41 +214,37 @@ def step(self, *args, **kwargs):
         sel.click(InfoBlock.element('Relationships', 'Images'))
 
 
-def get_all_providers(do_not_navigate=False):
+def get_all_providers():
     """Returns list of all providers"""
-    if not do_not_navigate:
-        navigate_to(CloudProvider, 'All')
-    providers = set([])
-    link_marker = "ems_cloud"
-    for page in paginator.pages():
-        for title in sel.elements("//div[@id='quadicon']/../../../tr/td/a[contains(@href,"
-                "'{}/show')]".format(link_marker)):
-            providers.add(sel.get_attribute(title, "title"))
-    return providers
-
-
-def discover(credential, cancel=False, d_type="Amazon"):
+    view = navigate_to(CloudProvider, 'All')
+    return [item.name for item in view.items.get_all(surf_pages=True)]
+
+
+def discover(credential, cancel=False, d_type="Amazon EC2"):
     """
     Discover cloud providers. Note: only starts discovery, doesn't
     wait for it to finish.
 
     Args:
       credential (cfme.base.credential.Credential):  Amazon discovery credentials.
       cancel (boolean):  Whether to cancel out of the discover UI.
+      d_type: provider name which will be discovered
     """
-    navigate_to(CloudProvider, 'Discover')
-    form_data = {'discover_select': d_type}
-    if credential:
-        form_data.update({'username': credential.principal,
-                          'password': credential.secret,
-                          'password_verify': credential.verify_secret})
-    fill(discover_form, form_data,
-         action=form_buttons.cancel if cancel else discover_form.start_button,
-         action_always=True)
+    view = navigate_to(CloudProvider, 'Discover')
+
+    view.fill({'discover_type': d_type,
+               'username': getattr(credential, 'principal', None),
+               'password': getattr(credential, 'secret', None),
+               'password_verify': getattr(credential, 'verify_secret', None)})
+    if cancel:
+        view.cancel.click()
+    else:
+        view.start.click()
```

Comment:
I believe some assertion(s) should be here, after the button click the view has to be changed. Also I think this function could return a view, but it can be done later.